README.md
# timewarp

[![crates.io](https://img.shields.io/crates/v/timewarp.svg)](https://crates.io/crates/timewarp)
[![crates.io](https://img.shields.io/crates/d/timewarp.svg)](https://crates.io/crates/timewarp)
[![Documentation](https://docs.rs/tolerance/badge.svg)](https://docs.rs/timewarp)

NLP library for parsing English and German natural language into dates and times. 
Leverages [pest](https://pest.rs) for parsing human readable-dates and times. 
Tries to be as lenient as possible. 

## Examples:

### Direct Input

To input a `2022-12-01` you can type:  `12/1/22`, `01.12.22`, `22-12-01`.

### Week

`2022-W52`, `2022W52`, `week 22-52` or `KW 22/52` are interpreted as an 
intervall `2022-12-26 <= x < 2023-01-02`. 

### Relative Dates

`yesterday`, `tomorrow`, etc. are calculated based of a given base.
`+4 weeks`, `-5 months`, `next friday`, `last thu` ... 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.4.0 (2023-12-06)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 1 commit contributed to the release.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Api streamlined. ([`8fe5460`](https://github.com/NORICS-net/timewarp/commit/8fe546054e072ea31c439285f582427bedd73e32))
</details>

## v0.3.2 (2023-12-04)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 7 commits contributed to the release.
 - 221 days passed between releases.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release timewarp v0.3.2 ([`3e5cd94`](https://github.com/NORICS-net/timewarp/commit/3e5cd94f29e307f2c44a872f6c0ab2068b2b6308))
    - Adjusting changelogs prior to release of timewarp v0.3.2 ([`1e5b99f`](https://github.com/NORICS-net/timewarp/commit/1e5b99fee40a5b5b24ceddc0977109a6a738aee4))
    - Adjusting changelogs prior to release of timewarp v0.3.2 ([`b8b431e`](https://github.com/NORICS-net/timewarp/commit/b8b431e8cfa051f33df7795e87f439ed1e66ad36))
    - Adjusting changelogs prior to release of timewarp v0.3.2 ([`5a7cf74`](https://github.com/NORICS-net/timewarp/commit/5a7cf74e3011daf4ee969801a1385c62357c8f3a))
    - CHANGELOG added. ([`cb40a34`](https://github.com/NORICS-net/timewarp/commit/cb40a342c157508712b93bb5e8920aa3e3960fbb))
    - Update README.md ([`a5e5f04`](https://github.com/NORICS-net/timewarp/commit/a5e5f046f3cb8f5fdc682c5f034adca319d64b5a))
    - Updated dependencies, better docu. ([`cfc7cf1`](https://github.com/NORICS-net/timewarp/commit/cfc7cf1976a643a75f9191d6038d2d8bca808daa))
</details>

## v0.3.1 (2023-04-27)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 1 commit contributed to the release.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Direction-enums renamed. ([`f97b926`](https://github.com/NORICS-net/timewarp/commit/f97b92679491f3c82f39058390538f236d9e2b06))
</details>

## v0.3.0 (2023-04-26)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 3 commits contributed to the release over the course of 18 calendar days.
 - 23 days passed between releases.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Iso8601-week dates and more added. ([`d186fcd`](https://github.com/NORICS-net/timewarp/commit/d186fcd9e24aab40422a7d3a4c083ade40aaa9e0))
    - Adding relative addition / subtraction '+5 [days|months|years]'. ([`ef72ae3`](https://github.com/NORICS-net/timewarp/commit/ef72ae3dce61f190fcff6e9324ab23463c3bf719))
    - DaySpan changed + docs improved. ([`5feb7ed`](https://github.com/NORICS-net/timewarp/commit/5feb7edec09aaa14ff169a4036a188985591fb8d))
</details>

## v0.2.0 (2023-04-02)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 4 commits contributed to the release over the course of 16 calendar days.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Month-names allowed. ([`eccfd05`](https://github.com/NORICS-net/timewarp/commit/eccfd0514aa2abb5e6f31626d9a910c95e6bd5b0))
    - Create .github/workflows/rust.yml ([`ec2c5e6`](https://github.com/NORICS-net/timewarp/commit/ec2c5e6b775cbe51975e9c0dcc2bc9bc944fb4a7))
    - First step: parses iso / de / en -dates. ([`c6f7a61`](https://github.com/NORICS-net/timewarp/commit/c6f7a61eab166b9f0eb2d5235e216824e1729a7a))
    - Initial commit ([`ad87fc3`](https://github.com/NORICS-net/timewarp/commit/ad87fc3ae32693f63be4acea3b635c81f9eb8a0c))
</details>



src/doy.rs
use crate::day_of_week::DayOfWeek;
use crate::error::parse_error;
use crate::month_of_year::Month;
use crate::DayOfWeek::{Fri, Sun, Thu};
use crate::TimeWarpError;
use std::cmp::Ordering;
use std::convert::TryFrom;
use std::fmt::{Debug, Display, Formatter};
use std::num::ParseIntError;
use std::ops::{Add, Sub};
use std::time::SystemTime;

/// Day Of Year. Helper-class to easily calculate dates.
#[must_use]
#[derive(Eq, PartialEq, Copy, Clone, Debug)]
pub struct Doy {
    pub year: i32,
    pub doy: i32,
}

impl Doy {
    pub const SECOND: u128 = 1000;
    pub const MINUTE: u128 = Self::SECOND * 60;
    pub const HOUR: u128 = Self::MINUTE * 60;
    pub const DAY: u128 = Self::HOUR * 24;
    pub const YEAR: u128 = Self::DAY * 365 + Self::HOUR * 6;

    /// returns the Doy representing today.
    pub fn today() -> Self {
        let millis = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_millis();
        Self::from_millis(millis)
    }

    /// converts milliseconds from POSIX time or Epoch time to Doy.
    pub fn from_millis(millis: u128) -> Self {
        let offset = millis % Self::YEAR;
        let year = 1970 + ((millis - offset) / Self::YEAR) as i32;
        let doy_offset = offset % Self::DAY;
        let doy = 1 + ((offset - doy_offset) / Self::DAY) as i32;
        Self { year, doy }
    }

    /// Creates a new Doy, by the give `dayOfYear` and the `year`.
    /// 1 = 1. Jan, 32 = 1. Feb, 0 = 31. Dec (year - 1)  
    pub fn new(doy: i32, year: i32) -> Self {
        if doy < 1 {
            return Self::new(365 + i32::from(Self::is_leapyear(year - 1)) + doy, year - 1);
        }
        let max_doy = 365 + i32::from(Self::is_leapyear(year));
        if doy > max_doy {
            Self::new(doy - max_doy, year + 1)
        } else {
            Self { year, doy }
        }
    }

    #[inline]
    fn day_per_month(year: i32) -> Vec<i32> {
        let leap = Self::is_leapyear(year) as i32;
        vec![31, 28 + leap, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    }

    /// Creates a Doy from `year`, `month` and `day`.
    ///
    /// # Panics
    /// panics if `month` is not in 1..12
    pub fn from_ymd(year: i32, month: i32, day: i32) -> Self {
        assert!(month > 0 && month < 13, "Month has to be in 1..12");
        let day_of_year = Self::day_per_month(year)
            .iter()
            .take(month as usize - 1)
            .sum::<i32>()
            + day;
        Self::new(day_of_year, year)
    }

    /// Creates a Doy for the Monday of the given week (iso 8601)
    ///
    /// # Panics
    /// panics if `week` is not in 1..53
    pub fn from_week(year: i32, week: i32) -> Self {
        assert!(week > 0 && week < 54, "Week has to be in 1..53");
        // weekday of 4th, Jan.
        let weekday = Self::new(4, year).day_of_week();
        let day_of_year = (week - 1) * 7
            + match weekday {
                Sun => -2,
                _ => Fri as i32 - weekday as i32,
            };
        Self::new(day_of_year, year)
    }

    /// Is the given `year` a leap-year?
    #[inline]
    pub fn is_leapyear(year: i32) -> bool {
        year % 4 == 0 && year % 100 != 0
    }

    /// Is this year a leap-year?
    pub fn leapyear(self) -> bool {
        Self::is_leapyear(self.year)
    }

    /// converts a *day of year* to `mmdd`.
    fn as_date(self) -> (i32, i32) {
        let mut doy = self.doy;
        let mut m = 1;
        for ds in Self::day_per_month(self.year) {
            if doy <= ds {
                return (m, doy);
            }
            m += 1;
            doy -= ds;
        }
        (-1, -1)
    }

    /// returns this doy in iso-format `yyyy-mm-dd`.
    pub fn as_iso_date(self) -> String {
        format!("{self:#}")
    }

    /// Day of Week
    #[inline]
    pub fn day_of_week(self) -> DayOfWeek {
        let y = self.year % 100;
        let y_off = y + (y / 4) + 6 - self.leapyear() as i32;
        DayOfWeek::from(y_off + self.doy)
    }

    /// The ISO 8601 Weeks start with Monday and end on Sunday. The first week of the year always
    /// contains January 4th. And the first Thursday is always in the first week of the year.
    ///
    /// returns the week in iso-8601-format: `yyyy`-W`ww`
    pub fn iso8601week(self) -> String {
        let dow = self.day_of_week();
        let thursday = match dow {
            Sun => self + Thu - 7, // Sunday => last day of ISO-week.
            _ => self + Thu - dow,
        };
        let kw = (thursday.doy + 6) / 7;
        format!("{}-W{kw:02}", thursday.year)
    }

    /// Returns the day of month.
    pub fn day_of_month(self) -> i32 {
        self.as_date().1
    }

    /// Returns just the `Month`.
    pub fn month(self) -> Month {
        Month::from(self.as_date().0)
    }
}

impl From<Doy> for String {
    fn from(doy: Doy) -> Self {
        doy.to_string()
    }
}

impl From<u128> for Doy {
    fn from(value: u128) -> Self {
        Doy::from_millis(value)
    }
}

impl Display for Doy {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        let (month, day) = self.as_date();
        let year = self.year;
        if f.alternate() {
            write!(f, "{year:04}-{month:02}-{day:02}")
        } else {
            write!(f, "{year:04}{month:02}{day:02}")
        }
    }
}

macro_rules! gen_calcs {
    ($($key:ident),+) => {
    $(
        impl Add<$key> for Doy {
            type Output = Doy;

            fn add(self, rhs: $key) -> Self::Output {
                Doy::new(self.doy + rhs as i32, self.year)
            }
        }

        impl Sub<$key> for Doy {
            type Output = Doy;

            fn sub(self, rhs: $key) -> Self::Output {
                Doy::new(self.doy - rhs as i32, self.year)
            }
        }
    )+
    }
}

gen_calcs!(i8, i16, i32, i64, u8, u16, u32, u64, DayOfWeek);

impl PartialOrd for Doy {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        let p = if self.lt(other) {
            Ordering::Less
        } else if self.gt(other) {
            Ordering::Greater
        } else {
            Ordering::Equal
        };
        Some(p)
    }

    fn lt(&self, other: &Self) -> bool {
        self.year < other.year || (self.year == other.year && self.doy < other.doy)
    }

    fn le(&self, other: &Self) -> bool {
        self.year < other.year || (self.year == other.year && self.doy <= other.doy)
    }

    fn gt(&self, other: &Self) -> bool {
        self.year > other.year || (self.year == other.year && self.doy > other.doy)
    }

    fn ge(&self, other: &Self) -> bool {
        self.year > other.year || (self.year == other.year && self.doy >= other.doy)
    }
}

impl TryFrom<&str> for Doy {
    type Error = TimeWarpError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        use std::str::FromStr;
        let err = |e: ParseIntError| -> Result<i32, TimeWarpError> {
            parse_error(format!("Error converting into number: '{value}'\n{e}",))
        };
        let y = i32::from_str(&value[0..4]).map_err(err)?;
        let (m, d) = if value.len() == 10 && &value[4..5] == "-" && &value[7..8] == "-" {
            (
                i32::from_str(&value[5..7]).map_err(err)?,
                i32::from_str(&value[8..10]).map_err(err)?,
            )
        } else if value.len() == 8 {
            (
                i32::from_str(&value[4..6]).map_err(err)?,
                i32::from_str(&value[6..8]).map_err(err)?,
            )
        } else {
            return parse_error(format!("Wrong date-format: '{value}'"));
        };
        if m < 1 || m > 12 {
            return parse_error(format!("Month out of range 0..12: '{m}'"));
        }
        let days_in_month = Self::day_per_month(y).as_slice()[(m - 1) as usize];
        if d < 1 || d > days_in_month {
            return parse_error(format!(
                "Days exceeded in month {m} '{d}' ({days_in_month})"
            ));
        }
        Ok(Self::from_ymd(y, m, d))
    }
}

/// A timespan in whole days.
///
#[derive(Debug, Eq, PartialEq)]
pub enum Tempus {
    Moment(Doy),
    Interval(Doy, Doy),
}

impl Tempus {
    /// The start-date of this `DaySpan` (inclusive).
    pub fn start(&self) -> Doy {
        match *self {
            Tempus::Moment(d) | Tempus::Interval(d, _) => d,
        }
    }

    /// The end-date of this `DaySpan` (exclusive).
    pub fn end(&self) -> Doy {
        match *self {
            Tempus::Moment(d) => d + 1,
            Tempus::Interval(_, e) => e,
        }
    }
}

#[cfg(test)]
mod should {
    use crate::day_of_week::DayOfWeek::*;
    use crate::doy::Doy;
    use crate::month_of_year::Month;
    use std::convert::TryFrom;

    #[test]
    fn try_from() {
        assert_eq!(
            "2018-01-01",
            Doy::try_from("2018-01-01").unwrap().as_iso_date()
        );
        assert!(Doy::try_from("2018-13-01").is_err());
        assert!(Doy::try_from("2018-02-29").is_err());
        assert!(Doy::try_from("20180431").is_err());
        assert!(Doy::try_from("2018/04/15").is_err());
    }

    #[test]
    fn from_week_of_year() {
        assert_eq!("2018-01-01", Doy::from_week(2018, 1).as_iso_date());
        assert_eq!("2018-12-31", Doy::from_week(2019, 1).as_iso_date());
        assert_eq!("2019-12-30", Doy::from_week(2020, 1).as_iso_date());
        assert_eq!("2021-01-04", Doy::from_week(2021, 1).as_iso_date());
        assert_eq!("2022-01-03", Doy::from_week(2022, 1).as_iso_date());
    }

    #[test]
    fn into_week_of_year() {
        // 1. Rule: 4th of January is always in W01
        assert_eq!("2018-W01", Doy::from_ymd(2018, 1, 4).iso8601week());
        assert_eq!("2019-W01", Doy::from_ymd(2019, 1, 4).iso8601week());
        assert_eq!("2020-W01", Doy::from_ymd(2020, 1, 4).iso8601week());
        assert_eq!("2021-W01", Doy::from_ymd(2021, 1, 4).iso8601week());
        assert_eq!("2022-W01", Doy::from_ymd(2022, 1, 4).iso8601week());
        assert_eq!("2023-W01", Doy::from_ymd(2023, 1, 4).iso8601week());
        assert_eq!("2026-W01", Doy::from_ymd(2026, 1, 4).iso8601week());

        assert_eq!("2018-W01", Doy::from_ymd(2018, 1, 1).iso8601week());
        assert_eq!("2019-W01", Doy::from_ymd(2019, 1, 1).iso8601week());
        assert_eq!("2020-W53", Doy::from_ymd(2021, 1, 1).iso8601week());
        assert_eq!("2021-W52", Doy::from_ymd(2022, 1, 1).iso8601week());

        assert_eq!("2018-W26", Doy::from_ymd(2018, 7, 1).iso8601week());
        assert_eq!("2019-W27", Doy::from_ymd(2019, 7, 1).iso8601week());
        assert_eq!("2020-W27", Doy::from_ymd(2020, 7, 1).iso8601week());
        assert_eq!("2021-W26", Doy::from_ymd(2021, 7, 1).iso8601week());
    }

    #[test]
    fn day_of_month() {
        let test = Doy::from_ymd(2018, 4, 13);
        assert_eq!(test.as_iso_date(), "2018-04-13");
        assert_eq!(test.month(), Month::Apr);

        let test = Doy::from_ymd(2018, 3, 6);
        assert_eq!(test.as_iso_date(), "2018-03-06");
        assert_eq!(test.month(), Month::Mar);
    }

    #[test]
    fn create_by_doy_year() {
        let proof = Doy::new(-7, 2020);
        let test = Doy::new(358, 2019);
        assert_eq!(test, proof);
        let proof = Doy::new(-1, 2020);
        assert_eq!("20191230", proof.to_string());
        let proof = Doy::new(-1, 2021);
        assert_eq!("20201230", proof.to_string());
    }

    #[test]
    fn return_leapyear() {
        assert!(Doy::new(1, 2020).leapyear());
        assert!(!Doy::new(1, 2018).leapyear());
        assert!(!Doy::new(1, 2000).leapyear());
    }

    #[test]
    fn convert_to_string() {
        assert_eq!("20201225", Doy::new(360, 2020).to_string());
        assert_eq!("20181225", Doy::new(359, 2018).to_string());
    }

    #[test]
    fn calc_day_of_week() {
        assert_eq!(Wed, Doy::new(31, 2018).day_of_week());
        assert_eq!(Thu, Doy::new(31, 2019).day_of_week());
        assert_eq!(Fri, Doy::new(31, 2020).day_of_week());
        // Wochentag vom 1. Weihnachtstag 25.12.
        assert_eq!(Tue, Doy::new(359, 2018).day_of_week());
        assert_eq!(Fri, Doy::new(360, 2020).day_of_week());
        assert_eq!(Sat, Doy::new(359, 2021).day_of_week());
    }

    #[test]
    fn create_via_try_from() {
        assert_eq!("20200229", Doy::from_ymd(2020, 2, 29).to_string());
        assert_eq!("19990814", Doy::from_ymd(1999, 8, 14).to_string());
        let d = "20240721";
        assert_eq!(d, &Doy::try_from(d).unwrap().to_string());
    }

    #[test]
    fn order_gt_or_lt() {
        let a = Doy::new(112, 2020);
        let b = Doy::new(225, 2020);
        let c = Doy::new(85, 2021);

        assert!(a < b);
        assert!(c > a);
        assert!(b < c);
        assert!(a >= a);
        assert!(b <= c);
    }

    #[test]
    fn add_i32() {
        let d = Doy::new(15, 2020) + 2;
        assert_eq!(Doy::new(17, 2020), d);
    }

    #[test]
    fn from_millis() {
        assert_eq!("20230317", Doy::from_millis(1679086777511).to_string());
        assert_eq!("20230101", Doy::from_millis(1672570315000).to_string());
        assert_eq!("20181231", Doy::from_millis(1546253515000).to_string());
    }
}


src/date_matcher.rs
use super::TimeWarpError;
use crate::day_of_week::DayOfWeek;
use crate::doy::{Doy, Tempus};
use crate::error::parse_error;
use crate::month_of_year::Month;
use pest::iterators::Pairs;
use pest::Parser;
use std::str::FromStr;

#[derive(Parser, Debug, Default)]
#[grammar = "date_matcher.pest"]
struct DateMatcher;

/// Designated use of the date.
#[derive(Eq, PartialEq, Debug, Copy, Clone)]
pub enum Direction {
    To,
    From,
}

#[allow(clippy::unnecessary_wraps)]
#[inline]
fn ok_moment(d: Doy) -> Result<Tempus, TimeWarpError> {
    Ok(Tempus::Moment(d))
}

fn correct_yyyy(yy: i32, relative: i32) -> i32 {
    if yy > 100 {
        return yy;
    }
    let offset = relative % 100;
    let base = relative - offset;
    if yy > offset + 50 {
        base - 100 + yy
    } else if yy < offset - 50 {
        base + 100 + yy
    } else {
        base + yy
    }
}

fn yy_mm_dd(pairs: Pairs<'_, Rule>, today: Doy) -> Result<Tempus, TimeWarpError> {
    let mut yy = today.year;
    let mut mm = 0;
    let mut dd = 0;
    for pair in pairs {
        match pair.as_rule() {
            Rule::yyyy => yy = correct_yyyy(i32::from_str(pair.as_str())?, today.year),
            Rule::mm => mm = i32::from_str(pair.as_str())?,
            Rule::dd => dd = i32::from_str(pair.as_str())?,
            _ => return parse_error(format!("No date. Found more than expected: {pair:?}")),
        };
    }
    ok_moment(Doy::from_ymd(yy, mm, dd))
}

fn date_long(pairs: Pairs<'_, Rule>, today: Doy) -> Result<Tempus, TimeWarpError> {
    let mut yy = today.year;
    let mut mm = 0;
    let mut dd = 0;
    for pair in pairs {
        match pair.as_rule() {
            Rule::yyyy => yy = correct_yyyy(i32::from_str(pair.as_str())?, today.year),
            Rule::month => {
                mm = Month::from_month(pair.into_inner().next().unwrap().as_rule()) as i32;
            }
            Rule::dd => dd = i32::from_str(pair.as_str())?,
            _ => return parse_error(format!("No long-date. Found more than expected: {pair:?}")),
        };
    }
    if yy < 100 {
        yy += 2000;
    }
    ok_moment(Doy::from_ymd(yy, mm, dd))
}

fn date_week(pairs: Pairs<'_, Rule>, today: Doy) -> Result<Tempus, TimeWarpError> {
    let mut yy = today.year;
    let mut kw = 0;
    for pair in pairs {
        match pair.as_rule() {
            Rule::yyyy => yy = correct_yyyy(i32::from_str(pair.as_str())?, today.year),
            Rule::kw => kw = i32::from_str(pair.as_str())?,
            _ => return parse_error(format!("No week-date. Found more than expected: {pair:?}")),
        }
    }
    let start = Doy::from_week(yy, kw);
    Ok(Tempus::Interval(start, start + 7))
}

pub fn date_matcher(
    today: Doy,
    direction: Direction,
    date: impl Into<String>,
) -> Result<Tempus, TimeWarpError> {
    let text = date.into();
    let mut amount = 0i32;
    let mut forwards = direction == Direction::To;
    for pair in DateMatcher::parse(Rule::date_matcher, &text)?
        .next()
        .unwrap()
        .into_inner()
    {
        match pair.as_rule() {
            Rule::date_iso | Rule::date_en | Rule::date_de => {
                return yy_mm_dd(pair.into_inner(), today)
            }
            Rule::date_long => return date_long(pair.into_inner(), today),
            Rule::date_kw => return date_week(pair.into_inner(), today),
            Rule::today => return ok_moment(today),
            Rule::yesterday => return ok_moment(today - 1),
            Rule::tomorrow => return ok_moment(today + 1),
            Rule::last => forwards = false,
            Rule::next => forwards = true,
            Rule::amount => amount = i32::from_str(pair.as_str())?,
            Rule::fore_last => {
                forwards = false;
                amount = 1;
            }
            Rule::after_next => {
                forwards = true;
                amount = 1;
            }
            Rule::day_of_week => {
                let wd_today = today.day_of_week();
                let target_wd =
                    DayOfWeek::from_day_of_week(pair.into_inner().next().unwrap().as_rule());
                let date = if forwards {
                    today + target_wd.days_before(wd_today) + amount * 7
                } else {
                    today - wd_today.days_before(target_wd) - amount * 7
                };
                return ok_moment(date);
            }
            Rule::month => {
                let month = Month::from_month(pair.into_inner().next().unwrap().as_rule());
                let date = find_rel_month(today, direction, forwards, month);
                return ok_moment(date);
            }
            Rule::timeunit => {
                return ok_moment(find_timeunit(
                    pair.into_inner().next().unwrap().as_rule(),
                    today,
                    amount,
                ))
            }
            _ => return parse_error(format!("date_matcher :: {pair:?}")),
        };
    }
    parse_error("Nothing found")
}

fn find_rel_month(today: Doy, direction: Direction, future: bool, target_month: Month) -> Doy {
    // if direction is EndTime add a Month
    let target_month = target_month + i32::from(direction == Direction::To);
    let today_m = today.month();
    let add = if target_month > today_m && !future {
        -1
    } else {
        i32::from(target_month < today_m && future)
    };
    Doy::from_ymd(today.year + add, target_month as i32, 1)
}

fn find_timeunit(rule: Rule, today: Doy, amount: i32) -> Doy {
    match rule {
        Rule::days => today + amount,
        Rule::months => {
            let mut m = today.month() as i32 + amount;
            let mut y = today.year;
            while m > 12 {
                m -= 12;
                y += 1;
            }
            while m < 1 {
                m += 12;
                y -= 1;
            }
            Doy::from_ymd(y, m, today.day_of_month())
        }
        Rule::years => Doy::new(today.doy, today.year + amount),
        _ => today,
    }
}

#[cfg(test)]
mod should {
    use crate::date_matcher;
    use crate::date_matcher::{correct_yyyy, find_rel_month};
    use crate::Direction::{From, To};
    use crate::Month::{Aug, Jan};
    use crate::{Doy, Tempus};

    #[test]
    fn adjust_yyyy() {
        assert_eq!(2023, correct_yyyy(2023, 2023));
        assert_eq!(2023, correct_yyyy(23, 2023));
        assert_eq!(2023, correct_yyyy(23, 1995));
        assert_eq!(1989, correct_yyyy(89, 2023));
        assert_eq!(2089, correct_yyyy(89, 2043));
    }

    #[test]
    fn find_relative_months() {
        // Fri 2023-03-17
        let today = Doy::from_ymd(2023, 3, 17);
        assert_eq!(Doy::new(1, 2023), find_rel_month(today, From, false, Jan));

        assert_eq!(
            Tempus::Moment(Doy::new(1, 2023)),
            date_matcher(today, From, "last january").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::new(1, 2024)),
            date_matcher(today, From, "next january").unwrap(),
        );
        assert_eq!(
            Doy::from_ymd(2023, 9, 1),
            find_rel_month(today, To, true, Aug)
        );
        assert_eq!(
            Doy::from_ymd(2022, 8, 1),
            find_rel_month(today, From, false, Aug)
        );
        assert_eq!(
            Doy::from_ymd(2022, 9, 1),
            find_rel_month(today, To, false, Aug)
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2024, 2, 1)),
            date_matcher(today, To, "next january").unwrap(),
        );
    }

    #[test]
    fn find_relative_dates() {
        // Fri 2023-03-17
        let today = Doy::from_ymd(2023, 3, 17);
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 13)),
            date_matcher(today, To, "last monday").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 14)),
            date_matcher(today, From, "tuesday").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 21)),
            date_matcher(today, To, "tuesday").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 16)),
            date_matcher(today, From, "letzten donnerstag").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 10)),
            date_matcher(today, To, "last friday").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 24)),
            date_matcher(today, To, "nächsten Fr").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 23)),
            date_matcher(today, To, "coming Thu").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 30)),
            date_matcher(today, To, "übernächsten Donnerstag").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 20)),
            date_matcher(today, To, "nächster Mo").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 6)),
            date_matcher(today, To, "vorletzter mo").unwrap(),
        );
    }

    #[test]
    fn find_yesterday() {
        let first_of_march = Doy::from_ymd(2023, 3, 1);
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 1)),
            date_matcher(first_of_march, To, "heute").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 2, 28)),
            date_matcher(first_of_march, To, "yesterday").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 2)),
            date_matcher(first_of_march, To, "morgen").unwrap(),
        );
    }

    #[test]
    fn adding_times() {
        // Fri 2023-03-17
        let today = Doy::from_ymd(2023, 3, 17);
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 22)),
            date_matcher(today, From, "+5 Tage").unwrap(),
        );
        let today = Doy::from_ymd(2023, 3, 17);
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2022, 3, 17)),
            date_matcher(today, From, "-1 year").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2022, 2, 17)),
            date_matcher(today, From, "-13 month").unwrap(),
        );
    }

    #[test]
    fn parse_date() {
        // Fri 2023-03-17
        let today = Doy::from_ymd(2023, 3, 17);
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 1, 22)),
            date_matcher(today, From, "22.01.2023").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 1, 22)),
            date_matcher(today, From, "22.1.23").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 1, 22)),
            date_matcher(today, From, "22.1.").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 16)),
            date_matcher(today, From, "3/16/2023").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 16)),
            date_matcher(today, From, "2023-03-16").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 16)),
            date_matcher(today, From, "    23-03-16  ").unwrap(),
        );

        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 16)),
            date_matcher(today, From, "16. Mär 2023").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 16)),
            date_matcher(today, From, "16. März 2023").unwrap(),
        );
        assert_eq!(
            Tempus::Moment(Doy::from_ymd(2023, 3, 16)),
            date_matcher(today, From, "March 16th 2023").unwrap(),
        );
    }

    #[test]
    fn parse_week() {
        let today = Doy::from_ymd(2023, 3, 17);

        assert_eq!(
            Tempus::Interval(Doy::from_ymd(2023, 3, 27), Doy::from_ymd(2023, 4, 3)),
            date_matcher(today, From, "2023-W13").unwrap(),
        );
        assert_eq!(
            Tempus::Interval(Doy::from_ymd(2020, 12, 21), Doy::from_ymd(2020, 12, 28)),
            date_matcher(today, From, "Woche 2020-52").unwrap(),
        );

        assert_eq!(
            Tempus::Interval(Doy::from_ymd(2020, 12, 21), Doy::from_ymd(2020, 12, 28)),
            date_matcher(today, From, "KW 20/52").unwrap(),
        );
    }
}
